features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/domain/ResourceRequestFactory.kt

```diff
@@ -0,0 +1,121 @@
+import com.datadog.android.sessionreplay.internal.recorder.SessionReplayResource
+import okhttp3.RequestBody
+import okio.Buffer
+import java.io.IOException
+import java.io.ObjectInputStream
+import java.io.StreamCorruptedException
+import java.lang.NullPointerException
+import java.util.Locale
+import java.util.UUID
+
+internal class ResourceRequestFactory(
+    internal val customEndpointUrl: String?,
+    private val rawEventsToResourcesMapper: RawEventsToResourcesMapper,
+    private val internalLogger: InternalLogger,
+    private val resourceRequestBodyFactory: ResourceRequestBodyFactory = ResourceRequestBodyFactory()
+) : RequestFactory {
+
+    override fun create(
+        context: DatadogContext,
+        batchData: List<RawBatchEvent>,
+        batchMetadata: ByteArray?
+    ): Request {
+        val resources = batchData.mapNotNull { rawBatchEvent ->
+            extractSessionReplayResource(rawBatchEvent)
+        }
+
+        if (resources.isEmpty()) {
+            @Suppress("ThrowingInternalException")
+            throw InvalidPayloadFormatException(INVALID_PAYLOAD_FORMAT_ERROR)
+        }
+
+        val requestBody = resourceRequestBodyFactory.create(resources, internalLogger)
+        return resolveRequest(context, requestBody)
+    }
+
+    private fun resolveRequest(context: DatadogContext, body: RequestBody): Request {
+        val bodyAsByteArray = extractByteArrayFromBody(body)
+        val requestId = UUID.randomUUID().toString()
+        val description = UPLOAD_DESCRIPTION
+        val headers = resolveHeaders(context, requestId)
+        val requestUrl = buildUrl(context)
+        return Request(
+            requestId,
+            description,
+            requestUrl,
+            headers,
+            body = bodyAsByteArray,
+            contentType = body.contentType().toString()
+        )
+    }
+
+    private fun extractByteArrayFromBody(body: RequestBody): ByteArray {
+        val buffer = Buffer()
+        @Suppress("UnsafeThirdPartyFunctionCall")
+        body.writeTo(buffer)
+        @Suppress("UnsafeThirdPartyFunctionCall")
+        return buffer.readByteArray()
+    }
+
+    private fun resolveHeaders(datadogContext: DatadogContext, requestId: String): Map<String, String> {
+        return mapOf(
+            RequestFactory.HEADER_API_KEY to datadogContext.clientToken,
+            RequestFactory.HEADER_EVP_ORIGIN to datadogContext.source,
+            RequestFactory.HEADER_EVP_ORIGIN_VERSION to datadogContext.sdkVersion,
+            RequestFactory.HEADER_REQUEST_ID to requestId
+        )
+    }
+
+    @SuppressWarnings("TooGenericExceptionCaught", "ThrowingInternalException")
+    private fun extractSessionReplayResource(rawBatchEvent: RawBatchEvent): SessionReplayResource? {
+        @Suppress("UnsafeThirdPartyFunctionCall")
+        return runCatching {
+            val ois = ObjectInputStream(rawBatchEvent.data.inputStream())
```

Comment:
why do we need to operate on the stream here? what is the rationale for that, given that all the bytes are already in the memory?